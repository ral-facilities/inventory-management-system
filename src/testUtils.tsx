import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
