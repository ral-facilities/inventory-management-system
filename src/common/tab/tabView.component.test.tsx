import { screen } from '@testing-library/react';

import userEvent from '@testing-library/user-event';
import { act } from 'react';
import { renderComponentWithRouterProvider } from '../../testUtils';
import TabView, { TabViewProps } from './tabView.component'; // Adjust import based on your directory structure

describe('TabView Component', () => {
  let props: TabViewProps<'tab1' | 'tab2' | 'tab3'>;

  const createView = (path: string) => {
    return renderComponentWithRouterProvider(
      <TabView {...props} />,
      'catalogueItem',
      path
    );
  };

  beforeEach(() => {
    props = {
      defaultTab: 'tab1',
      ariaLabelPrefix: 'test',
      tabData: [
        {
          value: 'tab1',
          icon: <div>icon 1</div>,
          component: <div>Content for Tab 1</div>,
        },
        {
          value: 'tab2',
          icon: <div>icon 2</div>,
          component: <div>Content for Tab 2</div>,
        },
        {
          value: 'tab3',
          icon: <div>icon 3</div>,
          component: <div>Content for Tab 3</div>,
        },
      ],
    };
  });

  it('renders correctly', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView('/catalogue/item/1').baseElement;
    });
    expect(baseElement).toMatchSnapshot();
  });
  it('renders correctly with default tab', () => {
    createView('/catalogue/item/1');

    expect(screen.getByText('Content for Tab 1')).toBeInTheDocument();
  });

  it('changes tabs', async () => {
    const { router } = createView('/catalogue/item/1');

    // Click on the second tab
    await userEvent.click(screen.getByText('tab2'));

    expect(await screen.findByText('Content for Tab 2')).toBeInTheDocument();
    expect(router.state.location.search).toBe('?tab=tab2');

    await userEvent.click(screen.getByText('tab3'));

    expect(await screen.findByText('Content for Tab 3')).toBeInTheDocument();
    expect(router.state.location.search).toBe('?tab=tab3');

    await userEvent.click(screen.getByText('tab1'));

    expect(await screen.findByText('Content for Tab 1')).toBeInTheDocument();
    expect(router.state.location.search).toBe('');
  });

  it('load tab from url', () => {
    createView('/catalogue/item/1?tab=tab2');

    expect(screen.getByText('Content for Tab 2')).toBeInTheDocument();
  });
});
