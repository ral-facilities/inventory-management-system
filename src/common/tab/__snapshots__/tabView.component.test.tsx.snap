// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabView Component > renders correctly 1`] = `
<body>
  <div>
    <div
      class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-i9v5tx-MuiGrid2-root"
    >
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
      >
        <div
          class="MuiTabs-root css-1ujnqem-MuiTabs-root"
        >
          <div
            class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-oqr85h"
            style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
          />
          <div
            class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
            style="margin-bottom: 0px;"
          >
            <div
              aria-label="test view tabs"
              class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
              role="tablist"
            >
              <button
                aria-controls="tab1-tabpanel"
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary Mui-selected css-12xzlyc-MuiButtonBase-root-MuiTab-root"
                id="tab1-tab"
                role="tab"
                tabindex="0"
                type="button"
              >
                <div
                  class="MuiTab-iconWrapper"
                >
                  icon 1
                </div>
                tab1
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="tab3-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary css-12xzlyc-MuiButtonBase-root-MuiTab-root"
                id="tab3-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="MuiTab-iconWrapper"
                >
                  icon 3
                </div>
                tab3
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="tab2-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary css-12xzlyc-MuiButtonBase-root-MuiTab-root"
                id="tab2-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <div
                  class="MuiTab-iconWrapper"
                >
                  icon 2
                </div>
                tab2
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Gallery-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary css-12xzlyc-MuiButtonBase-root-MuiTab-root"
                id="Gallery-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTab-iconWrapper css-i4bv87-MuiSvgIcon-root"
                  data-testid="CollectionsOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 4v12H8V4zm0-2H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2m-8.5 9.67 1.69 2.26 2.48-3.1L19 15H9zM2 6v14c0 1.1.9 2 2 2h14v-2H4V6z"
                  />
                </svg>
                Gallery
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-controls="Attachments-tabpanel"
                aria-selected="false"
                class="MuiButtonBase-root MuiTab-root MuiTab-labelIcon MuiTab-textColorPrimary css-12xzlyc-MuiButtonBase-root-MuiTab-root"
                id="Attachments-tab"
                role="tab"
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTab-iconWrapper css-i4bv87-MuiSvgIcon-root"
                  data-testid="AttachmentOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M18.5 16H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h12.5c1.38 0 2.5 1.12 2.5 2.5S20.88 13 19.5 13H9c-.55 0-1-.45-1-1s.45-1 1-1h9.5V9.5H9c-1.38 0-2.5 1.12-2.5 2.5s1.12 2.5 2.5 2.5h10.5c2.21 0 4-1.79 4-4s-1.79-4-4-4H7c-3.04 0-5.5 2.46-5.5 5.5s2.46 5.5 5.5 5.5h11.5z"
                  />
                </svg>
                Attachments
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
              style="left: 0px; width: 0px;"
            />
          </div>
        </div>
      </div>
      <div
        class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 css-16fnxy3-MuiGrid2-root"
      >
        <div
          aria-labelledby="tab1-tab"
          id="tab1-tabpanel"
          role="tabpanel"
          style="height: 100%;"
        >
          <div
            class="MuiBox-root css-10klw3m"
          >
            <div>
              Content for Tab 1
            </div>
          </div>
        </div>
        <div
          aria-labelledby="tab3-tab"
          hidden=""
          id="tab3-tabpanel"
          role="tabpanel"
          style="height: 100%;"
        />
        <div
          aria-labelledby="tab2-tab"
          hidden=""
          id="tab2-tabpanel"
          role="tabpanel"
          style="height: 100%;"
        />
        <div
          aria-labelledby="Gallery-tab"
          hidden=""
          id="Gallery-tabpanel"
          role="tabpanel"
          style="height: 100%;"
        />
        <div
          aria-labelledby="Attachments-tab"
          hidden=""
          id="Attachments-tabpanel"
          role="tabpanel"
          style="height: 100%;"
        />
      </div>
    </div>
  </div>
</body>
`;
